csrc/optimization/optimize_consecutive_cast.cpp

```diff
@@ -0,0 +1,110 @@
+    for (auto expr : fusion->exprs()) {
+      if (is_cast_op(expr)) {
+        bool mutated = false;
+        while (true) {
+          // in the loop, we just repetitively skip consecutive casts.
+          auto intermediate_cast = expr->input(0);
+          auto prev_expr = intermediate_cast->definition();
+          if (prev_expr != nullptr && is_cast_op(prev_expr)) {
+            auto original_dtype = prev_expr->input(0)->getDataType().value();
+            auto intermediate_dtype = intermediate_cast->getDataType().value();
+            auto out_dtype = expr->output(0)->getDataType().value();
+            // cases where skipping the intermediate cast is relatively safe,
+            // two conditions:
+            //   1. original_dtype is the same as out_dtype; or
+            //   2. we support direct cast from original_dtype to out_dtype.
+            // and
+            //   1. intermediate_dtype is the same type category as with
+            //   out_dtype; or
+            //   2. intermediate_dtype is more relaxed than out_dtype. e.g. a
+            //   floating point vs. integral;
+            if ((original_dtype == out_dtype ||
+                 cast_func_str({original_dtype, out_dtype}).has_value()) &&
+                ((isIntegralType(intermediate_dtype) &&
+                  isIntegralType(out_dtype)) ||
+                 (isFloatingPointType(intermediate_dtype) &&
+                  isFloatingPointType(out_dtype)) ||
+                 (isComplexType(intermediate_dtype) &&
+                  isComplexType(out_dtype)) ||
+                 (isFloatingPointType(intermediate_dtype) &&
+                  isIntegralType(out_dtype)))) {
+              expr = nvfuser::ir_utils::replaceValInExpr(
+                  expr, intermediate_cast, prev_expr->input(0));
+              mutated = true;
+            } else {
+              break;
+            }
+          } else {
+            break;
+          }
+        }
+
+        if (mutated) {
+          // quick short-wire to skip current cast node if it's trivially
+          // casting to the same type
+          if (expr->input(0)->getDataType().value() ==
+              expr->output(0)->getDataType().value()) {
+            replacement_map[expr->output(0)] = expr->input(0);
+            // NOTE: if current output is a fusion output, DCE won't kick in and
+            // we'll ended up with an illegal cast.
+            if (expr->output(0)->isFusionOutput()) {
+              fusion->replaceOutput(expr->output(0), expr->input(0));
+            }
+          }
+        }
+      }
+    }
+    if (!replacement_map.empty()) {
+      nvfuser::ir_utils::replaceValue(fusion, replacement_map);
+    }
+  }
+
+  std::string name() override {
+    return "ConsecutiveCastOptimization";
+  }
+
+  FusionPass func() override {
+    return runPass;
+  }
+};
+
+// triggering the ConsecutiveCastPass constructor to register the pass
+static ConsecutiveCastPass register_;
```

Comment:
Do we really need to use static variables like this? It seems convenient, for now, but It doesn't seem to work well as we add more passes.

Reply:
Agree'd on this one.
I'm just keeping something simple here to register a certain pass. I do tend to refactor this, because the direction is that we are going to assemble groups of passes. So each pass will be a functional call.

Do you prefer to have it refactored here already, or maybe I can do that in follow up PRs?

Reply:
Well, I think at least we should move this out of the static area. It should be fine for now, but, for example, if we added a new pass and again placed a pass object as a static variable, would there be a deterministic ordering of initializations? Suppose two passes don't specify any priority, but we still want to have a deterministic ordering for reproducibility. 

I understand that this PR is taking more time than expected, but I'd like to have this addressed in this PR.

Reply:
> if we added a new pass and again placed a pass object as a static variable, would there be a deterministic ordering of initializations? Suppose two passes don't specify any priority, but we still want to have a deterministic ordering for reproducibility.

That's a good point. The answer is no~~~ Unless there's a specified priority to dictate the order, it's just an arbitrary game. I was planning to refactor the registration. But I wouldn't mind doing a small portion in this PR.

I'll let you know when it's cleaned up. :bow: